File: library/src/main/java/me/xiaopan/sketch/Resize.java
/*
 * Copyright (C) 2013 Peng fei Pan <sky@xiaopan.me>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.sketch;

import android.widget.ImageView;

public class Resize implements ImageSize{
	private int width;
	private int height;

	private ImageView.ScaleType scaleType = ImageView.ScaleType.CENTER_CROP;

	public Resize(Resize sourceResize){
		this.width = sourceResize.width;
		this.height = sourceResize.height;
		this.scaleType = sourceResize.scaleType;
	}

	public Resize(int width, int height) {
		this.width = width;
		this.height = height;
	}

	public Resize(int width, int height, ImageView.ScaleType scaleType) {
		this(width, height);
		this.scaleType = scaleType;
	}

	public ImageView.ScaleType getScaleType() {
		return scaleType;
	}

	public void setScaleType(ImageView.ScaleType scaleType) {
		this.scaleType = scaleType;
	}

	@Override
	public int getWidth() {
		return width;
	}

	@Override
	public int getHeight() {
		return height;
	}

	@Override
	public void set(int width, int height) {
		this.width = width;
		this.height = height;
	}

	@Override
	public String getIdentifier(){
		return appendIdentifier(new StringBuilder()).toString();
	}

	@Override
	public StringBuilder appendIdentifier(StringBuilder builder){
		builder.append("Resize(");
		builder.append(width);
		builder.append("x");
		builder.append(height);
		if(scaleType != null){
			builder.append(":");
			builder.append(scaleType.name());
		}
		builder.append(")");
		return builder;
	}
}

File: sketch-core/src/main/java/com/github/panpf/sketch/request/Resize.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.request

import android.graphics.Bitmap
import android.widget.ImageView.ScaleType

data class Resize constructor(
    val width: Int,
    val height: Int,
    val mode: Mode = DEFAULT_MODE,
    val scaleType: ScaleType = DEFAULT_SCALE_TYPE,
    /**
     * Only applies to [Resize.Mode.THUMBNAIL_MODE]
     */
    val minAspectRatio: Float = DEFAULT_MIN_ASPECT_RATIO,
) {

    val cacheKey: String =
        "Resize(${width}x${height},${scaleType},${mode}${if (mode == Mode.THUMBNAIL_MODE) "-$minAspectRatio" else ""})"

    companion object {
        val DEFAULT_MODE: Mode = Mode.EXACTLY_SAME
        val DEFAULT_SCALE_TYPE: ScaleType = ScaleType.FIT_CENTER
        val DEFAULT_MIN_ASPECT_RATIO: Float = 1.5f
    }

    enum class Mode {
        /**
         * Even if the size of the original image is smaller than [Resize], you will get a [Bitmap] with the same size as [Resize]
         */
        EXACTLY_SAME,

        /**
         * The size of the new image will not be larger than [Resize], but the aspect ratio will be the same
         */
        ASPECT_RATIO_SAME,

        /**
         * If the difference between the aspect ratio of resize and the aspect ratio of image exceeds [minAspectRatio],
         * a portion of the original image will be captured based on the [width],[height],[scaleType] attributes.
         * This mode is suitable for displaying thumbnails of super-long images in nine squares
         */
        THUMBNAIL_MODE,
    }
}